Allow subprojects to be built for both the build and host machine (2nd part)

bonzini opened the pull request:
Building on #15395, actually add the functionality needed to build subproject for both machines. This is also propagated recursively to dependencies and subprojects of `native: true` subprojects.

Based on (and supersedes) #12258. Other than rebasing, the main logic changes are:

* reimplemented in the interpreter by rewriting the `native` and `install` keyword arguments. This ensures more easily that all created objects refer to the right machine; build targets check that they are for the build machine upon creation, and other objects then rely on the normal detection of mixed machines. It also simplifies the code, for exampe fixing project arguments
* rewritten the handling of the subproject stack, keeping a single stack but storing `(name, for_machine)` tuples in it. This makes it possible to present chains of subprojects more clearly as `a-0.1:b-2.0 (build):c-3.0`. Note that in such a scenario `c-3.0` is technically built for the *host* machine of `b-2.0 (build)` (which is the build machine of `a-0.1`). This is handled correctly by `copy_for_build_machine()` and `PerMachine`, but it messed up the old subproject stack handling.
* postconf scripts are not blocked. dist and postconf scripts should receive the subproject's roots, which is a separate change. This can be handled separately after #15395 is merged.

For now I've tested this on QEMU, by adding `native: true` to all subprojects linked to Rust procedural macro crates (such as `proc_macro2`, `quote`, `syn`) and removing it from the `dependency` and `static_library` invocations. All crates can find their dependencies correctly and there are no warnings about cross compilation breakage, including those that are only depended upon indirectly.

Fixes: #8043
Fixes: #8944

bonzini committed:
interpreter: block install_* commands for build subprojects

The object however has to be returned anyway.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
interpreter, cmake: add native argument to subproject()

Extracted from a patch by Dylan Baker <dylan@pnwbakers.com>

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

dnicolodi commented on docs/markdown/snippets/subproject_and_override_program_native.md:
```diff
@@ -0,0 +1,6 @@
+
+Subprojects may now be built for the host or build machine (or both). This means
+that build time dependencies can be built for the matching running the build in a
+cross compile setup. When a subproject is run for the build machine it will act
```
Something is off in this sentence.

bonzini committed:
Revert "build: move find_overrides away"

This reverts commit 8842bd44a703cf4da3a17a337c8488204c438844.
It is incorrect to keep find_overrides in Interpreter because it
won't be rolled back properly if a non-required subproject fails.

bonzini committed:
build: simplify rel_string_list

os.path.normpath will remove a trailing "/" or "/.", thus undoing the
effect of os.path.normpath(os.path.join(self.curdir, idir)) if idir
is "" or ".".  As such, expdir will always be the same as self.curdir;
remove it.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
codegen: do not use Programs.for_machine

Program.for_machine is used, when building the command line for executable or
tests, to possibly apply an exe_wrapper to LocalProgram instances.

However, mesonbuild/modules/codegen.py was incorrectly using flex's
for_machine field to look up a compiler.  When checking for FlexLexer.h,
the code needs the compiler for the host machine (where the generated
C/C++ code will be compiled), not the build machine (where flex runs).

Track which machine the generated output is for, and use it as needed.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
interpreter: apply machine map when decorator can't be used

There are a few rare cases where it is not possible to use
@apply_machine_map_to_kwargs:

- add_project_*_arguments, because in build-only subprojects
  `native: true` must be discarded

- add_languages, where absent `native` is not the same as
  `native: false`

- build targets (at least for now; this may change later if
  the interface to build is not anymore build/host/target but
  the machine name), so that the transformation is done in
  def build_target()

- modules that do not yet support a `native` argument

Handle them.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
build: differentiate build directory according to a subproject's build machine

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

dcbaker committed:
tests for native subprojects

bonzini committed:
interpreter, rust: add native argument to workspace object

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini commented:
@dcbaker, can you take a look once the interpreter typing PR goes in?

bonzini committed:
build, interpreter: make project-related fields per machine

This is a first step towards making it possible to allow subprojects to
be built for both the build and host machine.  Even though the concept
of a build-machine subproject doesn't exist yet, prepare for it by
making the list of (sub)projects per machine, and storing the machine
in the interpreter's subproject stack.

Extracted from a patch by Dylan Baker <dylan@pnwbakers.com>

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
environment, build: store which machines are in use

An environment does not necessarily specify all the machines which
are available: both build and target machines can default to host.
And ultimately, what defines if a machine is "native" is whether the
build machine resolves to it.

Place all the information on how to go from machine to its defaulted
definition into a new MachineMap object.  Ultimately this has two
purposes:

- allowing the user to keep reasoning on build/host/target if
  desirable, while letting environment operate on N+1 machines (one
  native and N cross)

- removing PerMachine and PerMachineDefaultable, in favor of using
  dict[str, ...] (possibly with MachineChoice as a newtype for the key)
  anywhere except in the interpreter.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
utils: add indexing to PerThreeMachine

PerThreeMachine does not allow accessing via the [] operator
(__getitem__/__setitem__).  Allow that, using a new enum as
the index type.  ThreeMachineChoice overlaps MachineChoice
enough that the accessors can take either enum.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
interpreter: allow subprojects to be requested for the build machine

The current situation in Meson is that subprojects are always built for
the host machine. This means that while cross compiling, a tool that
must be run on the build machine must have all of its dependencies met
by the system (pkg-config, cmake, config-tool, etc); but dependencies
for the host machine can be built. With this patch subprojects may also
be built for the build machine.

Nothing should ever be installed from a subproject for the build
machine, while cross compiling. However, if we are not cross compiling
they are installed, since Meson can't really tell the difference between
a host and build dependency in that case.

Extracted from a patch by Dylan Baker <dylan@pnwbakers.com>

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
interpreter: add native argument to override_find_program()

Co-authored-by: Dylan Baker <dylan@pnwbakers.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
tests: add test that native subprojects have different introspection ids

bonzini committed:
build: replace SubProject with BuildProject for target

The SubProject alone will not be a unique key into build.projects.
Just store the object itself, as it allows retrieving the key.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
test cases: expand test for nonexisting property

Cover various combination of the native keyword argument and whether the
property is in a native or cross file.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
interpreter: introduce apply_machine_map

Introduce a decorator that applies the transformation in
self.build.machine_map to kwargs['native'].  In the future
this will also handle conversion of kwargs['native'] to
kwargs['for_machine'].

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
build: keep host/build separate when merging from a build-only subproject

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
build: add subproject to BuildProject

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
pkgconfig: do not map machine in PkgConfigInterface._cli

It is only ever used with MachineChoice.HOST, so it does nothing.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
pkgconfig: do not map machine in PkgConfigInterface.instance

It is only ever called from _cli (which does not need to
map the MachineChoice) or PkgConfigDependency (which
is created - via DependencyFallbacksHolder - from
DependencyObjectKWs that already have a properly
mapped machine.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini commented:
Compared to earlier versions, this introduces a new concept, the "machine map", and uses it to translate `kwargs['native']` within build-only subprojects. The concept is generic, and in fact there's no mention of anything being "bulid-only" in the code.

As mentioned in #15895, this is a first step towards isolating build/host/target machines (and `native: true/false`) inside the interpreter, and making MachineChoice a newtype on str. This way the Environment, Build and Backend would be able to use arbitrary machines, and PerMachine would be replaced by a `dict[MachineChoice, ...]`. (The machine map would probably move from `Build` to `BuildProject`).

Within the interpreter, the new decorator `apply_machine_map` would translate `kwargs['native']` into `kwargs['for_machine']`. Since the interpreter would not use `kwargs['native']` at all, there would be no risk of leaving out the decorator. Note that it is already possible to introduce `kwargs['for_machine']` and to make `kwargs['native']` a bool, alleviating the worry that one forgets `apply_machine_map`, but I didn't do it to avoid further churn.

dcbaker commented:
Your first commit is running up against a real problem we have in cross files. Sometimes host machines binaries are binaries that "run on the host machine" and sometimes they're binaries that "generate code for the host machine" This is particularly important for flex because it's 100% possible that you need two different flex/lex implementations when cross compiling from Windows, you need win_flex which generates Windows compatible code for the build machine, and you need !win_flex to generate Unix compatible code for the host machine. The assumption of the codegen module is that flex is a "generates code for X machine" type tool.

We probably need to separate those two concepts into something like a `[toolchain]` and `[binaries]` section, where `[binaries]` are those that run on the machine, and `[toolchain]` are the tools that generate for the machine. We would then be able to assume outright that anything in the `[toolchain]` section does not need an exe-wrapper, and anything in the `[binaries]` section must have an exe wrapper.

bonzini commented:
I agree it's a real problem, but I am not sure there has to be a field in Program.

The solution I have in mind there is that, in the overall replacement of "native", `find_program`'s `for_machine` argument would represent the machine for which the program generates code. Therefore the program would be sought in the corresponding machine file, and it would also be passed to a hypothetical fallback subproject as the subproject's `target_machine`.

I haven't thought about binaries vs. toolchain, generally I have always assumed that binaries are for the build machine and used properties for host binaries (including not using find_program). But I think [binaries] has too much backwards compatibility baggage, so it would have to stay reserved for build machine binaries.

dcbaker commented:
I still don't understand the change to the codegen module. It seems like we're just doing extra work to get the exact same information, since in a cross compilation setup you have to specify `flex` and `bison` in your cross file, so the `program.for_machine` will always be correct.

bonzini commented:
Finally I found the comment that describes it:

```python
    @classmethod
    def from_bin_list(cls, env: 'Environment', for_machine: MachineChoice, name: str) -> 'ExternalProgram':
        # This is not the static `for_machine` in this class, which represents
        # that the binary always runs on the build platform. (Its host platform
        # is our build platform.) Some external programs have a target platform,
        # and that is what we are specifying here.
        command = env.lookup_binary_entry(for_machine, name)
        if command is None:
            return NonExistingExternalProgram()
        return cls.from_entry(name, command)
```

In other words `for_machine` is defined as the machine *on which the binary runs*. So, as the commit message says, flex's `for_machine` will always be `BUILD` and `comp = self.interpreter.environment.coredata.compilers[for_machine]['cpp']` looks up the wrong machine.

dcbaker commented:
Okay, that makes sense.

That patch is probably suitable for backport to 1.11.X then?

bonzini commented:
Woohoo, thank you so much for the review. I hope the rationale for MachineMap made sense to you.

Yes, I'll extract that patch into a separate PR and then merge this on top (tomorrow).

dcbaker commented:
It does, I'm a little sad to see `PerMachine` on the way out, it's a cool littel piece of code and has been very ergonomic to work with. But thus is life :)
